### PR TITLE
Add [experimental] methods for writing arrays and dataframes

### PR DESCRIPTION
**Pull request by jmaruland**
NOTE:
The methods that were added to node.py and context.py are intended to be experimental. We are testing approaches to make tiled receive data and put it in a node in the tree. This might be subject to changes in the future. The idea behind this approach is to make it as easy as possible to the user to use a write command  for tiled while the data validation, data storage and management happens internally and with no complicated or convoluted steps for the user.

### CODE REVIEW

File: tiled/client/node.py
Comment (danielballan) [inline tiled/client/node.py:552]: On second thought, I think `write_array` (matching `read()`) would be a better name here. From the user's point of view, they aren't _creating_ the array (they already have the array) they are writing it into tiled's storage.

File: tiled/client/node.py
Comment (danielballan) [inline tiled/client/node.py:585]: Same here: `write_dataframe`.

File: tiled/client/node.py
Comment (danielballan) [inline tiled/client/node.py:581]: If we write to a node like `/analyzed` this should go to `/node/metadata/analyzed/{uid}`. This example is relevant:

https://github.com/bluesky/tiled/blob/16478d9e55b6f1f401b98d9fe042e3d0ed4d7f80/tiled/client/array.py#L49-L53

File: tiled/client/node.py
Comment (danielballan) [inline tiled/client/node.py:583]: Same issue here about including the path to this node.

File: tiled/client/node.py
Comment (danielballan) [inline tiled/client/node.py:624]: Same issue here about including the path to this node.

File: tiled/client/node.py
Comment (danielballan) [inline tiled/client/node.py:629]: Same issue here about including the path to this node.

File: tiled/client/node.py
Comment (danielballan) [inline tiled/client/node.py:628]: Does this need to be Arrow, not CSV? The `serialize_arrow` function would be useful here.

**Review by danielballan — CHANGES_REQUESTED** [review]

File: tiled/adapters/tiff.py
Comment (danielballan) [inline tiled/adapters/tiff.py:219]: This reads the entire sequence (i.e. all the TIFF files) and then slices out the part we want. The existing code reads only exactly the data we need, which is of course faster. I am sure that tifffile still supports this somehow, but it's not immediately clear how. Let's address this tifffile change in a separate PR.

File: tiled/client/node.py
Comment (danielballan) [inline tiled/client/node.py:628]: Why base64 encode?

**Review by danielballan — CHANGES_REQUESTED** [review]

File: tiled/adapters/tiff.py
Comment (jmaruland) [reply tiled/adapters/tiff.py]: Fixed in https://github.com/bluesky/tiled/pull/206

**Review by danielballan — APPROVED** [review]

File: tiled/client/node.py
Comment (kleinhenz) [inline tiled/client/node.py:596]: Should we send `npy` instead? The metadata would be redundant with the micro/macro structure but that is also true for arrow.

Comment (kleinhenz) [reply tiled/client/node.py:596]: Should there be a general mechanism for providing a serializer to turn your object into bytes? For reading we allow you to request data in many different formats. Maybe the same should be true for writing?

**Review by danielballan — CHANGES_REQUESTED** [review]
Taking a closer look at this I have more questions before we merge. Will comment in detail later.

File: tiled/client/node.py
Comment (danielballan) [reply tiled/client/node.py]: As a wire format, I think the server should accept just the array buffer because that is a more cross-language concept, where the `npy` format (with some dtype header in there) is a Python-specific thing.

I think it would be sensible for the server to _persist_ the data as npy though.

**Review by danielballan — APPROVED** [review]
We resolved our concerns over a phone call. Will merge once @jmaruland confirms that this can integrate with his experimental databroker server extension.

**Pull request merged**
